README.md
phpspec-code-coverage
=====================
[![Latest Stable Version](https://img.shields.io/packagist/v/friends-of-phpspec/phpspec-code-coverage.svg?style=flat-square)](https://packagist.org/packages/friends-of-phpspec/phpspec-code-coverage)
[![Total Downloads](https://img.shields.io/packagist/dt/friends-of-phpspec/phpspec-code-coverage.svg?style=flat-square)](https://packagist.org/packages/friends-of-phpspec/phpspec-code-coverage)
[![Scrutinizer Code
Quality](https://img.shields.io/scrutinizer/quality/g/friends-of-phpspec/phpspec-code-coverage/master?style=flat-square)](https://scrutinizer-ci.com/g/friends-of-phpspec/phpspec-code-coverage/?branch=master)
[![Travis](https://img.shields.io/travis/com/friends-of-phpspec/phpspec-code-coverage.svg?style=flat-square)](https://travis-ci.org/friends-of-phpspec/phpspec-code-coverage)

[phpspec-code-coverage][0] is a [PhpSpec][2] extension that generates Code
Coverage reports for [PhpSpec][2] tests.

Generating Code Coverage reports allows you to to analyze which parts of your
codebase are tested and how well. However, Code Coverage alone should NOT be
used as a single metric defining how good your tests are.

![phpspec-code-coverage console report](https://i.imgur.com/BU10ZAV.png)
![phpspec-code-coverage HTML report](https://i.imgur.com/6xACR1g.png)

## Requirements

- PHP 7+ (for [PhpSpec][2] v4+) or PHP 5.6+ (for [PhpSpec][2] v3)
- [Xdebug][3], [phpdbg][4] or [pcov][6] extension enabled (PHP 7+ is required for code
  generation to work with [phpdbg][4]).

## Change Log

Please see [CHANGELOG.md](CHANGELOG.md) for information on recent changes.

## Install

Install this package as a development dependency in your project:

    $ composer require --dev friends-of-phpspec/phpspec-code-coverage

Enable extension by editing `phpspec.yml` of your project:

``` yaml
extensions:
  FriendsOfPhpSpec\PhpSpec\CodeCoverage\CodeCoverageExtension: ~
```

This will sufficient to enable Code Coverage generation by using defaults
provided by the extension. This extension supports various [configuration
options](#Options). For a fully annotated example configuration
file check [Configuration section](#Configuration).

## Usage

If you execute `phpspec run` command, you will see code coverage generated in `coverage` directory (in `html` format):

    $ bin/phpspec run

**Note!** When generating Code Coverage reports make sure PHP processes run via
CLI are not memory limited (i.e. `memory_limit` set to `-1` in
`/etc/php/cli/php.ini`).

### Running with phpdbg

This extension now supports [phpdbg][4], which results in faster execution when
using more recent versions of PHP. Run `phpspec` via [phpdbg][4]:

    $ phpdbg -qrr phpspec run

**Note!** PHP 7+ is required for code generation to work with [phpdbg][4].

## Configuration

You can see fully annotated `phpspec.yml` example file below, which can be used
as a starting point to further customize the defaults of the extension. The
configuration file below has all of the [Configuration Options](#Options).

```yaml
# phpspec.yml
# ...
extensions:
  # ... other extensions ...
  # friends-of-phpspec/phpspec-code-coverage
  FriendsOfPhpSpec\PhpSpec\CodeCoverage\CodeCoverageExtension:
    # Specify a list of formats in which code coverage report should be
    # generated.
    # Default: [html]
    format:
      - text
      - html
      #- clover
      #- php
      #- xml
    #
    # Specify output file/directory where code coverage report will be
    # generated. You can configure different output file/directory per
    # enabled format.
    # Default: coverage
    output:
      html: coverage
      #clover: coverage.xml
      #php: coverage.php
      #xml: coverage
    #
    # Should uncovered files be included in the reports?
    # Default: true
    #show_uncovered_files: true
    #
    # Set lower upper bound for code coverage
    # Default: 35
    #lower_upper_bound: 35
    #
    # Set high lower bound for code coverage
    # Default: 70
    #high_lower_bound: 70
    #
    # Whilelist directories for which code generation should be done
    # Default: [src, lib]
    #
    whitelist:
      - src
      - lib
    #
    # Whiltelist files for which code generation should be done
    # Default: empty
    #whilelist_files:
      #- app/bootstrap.php
      #- web/index.php
    #
    # Blacklist directories for which code generation should NOT be done
    #blacklist:
      #- src/legacy
    #
    # Blacklist files for which code generation should NOT be done
    #blacklist_files:
      #- lib/bootstrap.php
```

### Options

* `format` (optional) a list of formats in which code coverage should be
  generated. Can be one or many of: `clover`, `php`, `text`, `html`, `xml` (default
  `html`)
  **Note**: When using `clover` format option, you have to configure specific
  `output` file for the `clover` format (see below).
* `output` (optional) sets an output file/directory where specific code
  coverage format will be generated. If you configure multiple formats, takes
  a hash of `format:output` (e.g. `clover:coverage.xml`) (default `coverage`)
* `show_uncovered_files` (optional) for including uncovered files in coverage
  reports (default `true`)
* `lower_upper_bound` (optional) sets lower upper bound for code coverage
  (default `35`).
* `high_lower_bound` (optional) sets high lower bound for code coverage
  (default `70`)
* `whitelist` takes an array of directories to whitelist (default: `lib`,
  `src`).
* `whitelist_files` takes an array of files to whitelist (default: none).
* `blacklist` takes an array of directories to blacklist (default: `test,
  vendor, spec`)
* `blacklist_files` takes an array of files to blacklist

## Authors

Copyright (c) 2017-2018 ek9 <dev@ek9.co> (https://ek9.co).

Copyright (c) 2013-2016 Henrik Bjornskov, for portions of code from
[henrikbjorn/phpspec-code-coverage][1] project.

## License

Licensed under [MIT License](LICENSE).

[0]: https://github.com/friends-of-phpspec/phpspec-code-coverage
[1]: https://github.com/henrikbjorn/PhpSpecCodeCoverageExtension
[2]: http://www.phpspec.net/en/stable
[3]: https://xdebug.org/
[4]: https://github.com/krakjoe/phpdbg
[5]: https://github.com/leanphp/phpspec-code-coverage
[6]: https://github.com/krakjoe/pcov


CHANGELOG.md
# CHANGELOG

All notable changes to [https://github.com/friends-of-phpspec/phpspec-code-coverage][0] package will be
documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [4.3.2] - 2019-11-12

- Allow compatibility with PHPSpec v6

## [4.3.1] - 2019-10-01

- Update root namespace to `FriendsOfPhpSpec` but maintain a class alias to the
  old one `LeanPHP`. The `LeanPHP` root namespace is now deprecated and will be
  removed in v5.

## [4.3.0] - 2019-10-01

**Note!** This version mark the new home of the project. It was forked from
`leanphp/phpspec-code-coverage` project and rebranded as `friends-of-phpspec/phpspec-code-coverage`.
Find details here : https://github.com/friends-of-phpspec/phpspec-code-coverage

- PHPSpec 5 support

## [4.2.2] - 2018-03-22

- Bugfix: loosen up `phpunit/php-code-coverage` from `^6.0` to `^5.0||^6.0`
- PHP 7.1 specific code improvements

## [4.2.1] - 2018-03-21

- Integrate fixes from 4.1.2 (`--no-coverage` option bugfix)
- Cleanup development dependencies
- Minor code improvements

## [4.2.0] - 2018-03-19

- Updated `phpunit/php-code-coverage` dependency from `~5.0` to `~6.0`.
- Updated PHP requirement from `^7.0` to `^7.1`
- Updated `phpspec/phpspec` dependency from `~4.0` to `^4.2`

## [4.1.2] - 2018-03-21

- Fix `--no-coverage` option introducing errors when running non `run` commands.
- `--no-coverage` option is now available to all phpspec commands (not only
  `run`). (#30)

## [4.1.1] - 2018-03-19

- Added `--no-coverage` option which can skip code coverage generation during
  PhpSpec test run.

## [4.1.0] - 2018-03-17

- `phpunit/php-code-coverage` dependency version requirement has been updated
  from `~4.0|~5.0` to `~5.0` as we do not support version `4.0` anymore.
- Updated README with information regarding `memory_limit` when generating code
  coverage reports.
- PHP 7.2 has been added to travis test matrix

## [4.0.0] - 2017-10-17 - PhpSpec v4 / PHP 7+ release

This release adds support for PhpSpec v4. As a result of this, PHP requirement
has also been updated to PHP 7+.

- Added PhpSpec4 support #10
- Extension requires PHP7+ (due to PhpSpec v4 depending on it) #10

## [3.1.1] - 2017-10-17 - Maintenance release for PhpSpec v3

- PHPSpec version is now included when generating XML report #12
- Added example configuration options for generating XML report #12
- Minor cleanup in export-ignores. Should result in cleaner dist install #8

## [3.1.0] - 2017-02-21 (backported master branch on 2017-02-21)

**Note!** This is last backported release which pulls all the changes from the
`master` branch of `henrikbjorn/phpspec-code-coverage` project and releases it as
`3.1.0`.

- Add support for `php-code-coverage` v5.

## [3.0.1] - 2017-02-14 (backported v3.0.1, original release on 2016-08-02)

- Require PhpSpec3
- Require PHP 5.6+ / PHP 7.0+

## [2.1.0] - 2017-02-12 (backported v2.1.0, original release on 2016-05-05)

**Note!** v2.1.0 is final release to support PhpSpec2.

- Added PHP 7 support
- Added `phpdbg` extension support (alternative to `xdebug`)
- Updated `blacklist` to include `test` directory by default
- Updated `text` output to use coloring by default.

## [1.0.1] - 2017-02-11 (backported v1.0.1, original release on 2014-12-11)

**Note!** This version is a direct backport of `1.0.1` of
[henrikbjorn/phpspec-code-coverage][1] package with updated namespaces to work
as [leanphp/phpspec-code-coverage][0].

- PHP `>=5.3`
- PhpSpec `~2.0`
- Xdebug `>=2.1.4`
- Supports Code Coverage generation in `html`, `clover`, `php` and `txt`
  formats.
- Supports per-format `output` directory configuration (e.g.
  `clover:clover.xml`)
- Supports configuring inclusion of uncovered files in code coverage reports.
- Supports configuring lower upper and higher lower bounds for code coverage
  reports.
- Supports configuring a whiltelist of directories to be included in code
  coverage report (`whilelist` option).
- Supports configuring a whiltelist of files to be included in code coverage
  reports (`whitelist_files` option).
- Support configuring a blacklist of directories to be excluded from code
  coverage reports (`blacklist` option).
- Support configuring a blacklist of files to be excluded from code coverage
  reports (`blacklist_files` option).

[4.3.2]: https://github.com/friends-of-phpspec/phpspec-code-coverage/releases/tag/v4.3.2
[4.3.1]: https://github.com/friends-of-phpspec/phpspec-code-coverage/releases/tag/v4.3.1
[4.3.0]: https://github.com/friends-of-phpspec/phpspec-code-coverage/releases/tag/v4.3.0
[4.2.2]: https://github.com/friends-of-phpspec/phpspec-code-coverage/releases/tag/v4.2.2
[4.2.1]: https://github.com/friends-of-phpspec/phpspec-code-coverage/releases/tag/v4.2.1
[4.2.0]: https://github.com/friends-of-phpspec/phpspec-code-coverage/releases/tag/v4.2.0
[4.1.2]: https://github.com/friends-of-phpspec/phpspec-code-coverage/releases/tag/v4.1.2
[4.1.1]: https://github.com/friends-of-phpspec/phpspec-code-coverage/releases/tag/v4.1.1
[4.1.0]: https://github.com/friends-of-phpspec/phpspec-code-coverage/releases/tag/v4.1.0
[4.0.0]: https://github.com/friends-of-phpspec/phpspec-code-coverage/releases/tag/v4.0.0
[3.1.1]: https://github.com/friends-of-phpspec/phpspec-code-coverage/releases/tag/v3.1.1
[3.1.0]: https://github.com/friends-of-phpspec/phpspec-code-coverage/releases/tag/v3.1.0
[3.0.1]: https://github.com/friends-of-phpspec/phpspec-code-coverage/releases/tag/v3.0.1
[2.1.0]: https://github.com/friends-of-phpspec/phpspec-code-coverage/releases/tag/v2.1.0
[1.0.1]: https://github.com/friends-of-phpspec/phpspec-code-coverage/releases/tag/v1.0.1

[0]: https://github.com/friends-of-phpspec/phpspec-code-coverage
[1]: https://github.com/henrikbjorn/PhpSpecCodeCoverageExtension


spec/CodeCoverageExtensionSpec.php
<?php

declare(strict_types=1);

namespace spec\FriendsOfPhpSpec\PhpSpec\CodeCoverage;

use Exception;
use FriendsOfPhpSpec\PhpSpec\CodeCoverage\CodeCoverageExtension;
use PhpSpec\ObjectBehavior;
use PhpSpec\ServiceContainer\IndexedServiceContainer;

/**
 * @author Henrik Bjornskov
 */
class CodeCoverageExtensionSpec extends ObjectBehavior
{
    public function it_is_initializable(): void
    {
        $this->shouldHaveType(CodeCoverageExtension::class);
    }

    public function it_should_transform_format_into_array(): void
    {
        $container = new IndexedServiceContainer();
        $container->setParam('code_coverage', ['format' => 'html']);
        $this->load($container);

        $options = $container->get('code_coverage.options');

        if ($options['format'] !== ['html']) {
            throw new Exception('Default format is not transformed to an array');
        }
    }

    public function it_should_use_html_format_by_default(): void
    {
        $container = new IndexedServiceContainer();
        $this->load($container, []);

        $options = $container->get('code_coverage.options');

        if ($options['format'] !== ['html']) {
            throw new Exception('Default format is not html');
        }
    }

    public function it_should_use_singular_output(): void
    {
        $container = new IndexedServiceContainer();
        $container->setParam('code_coverage', ['output' => 'test', 'format' => 'foo']);
        $this->load($container);

        $options = $container->get('code_coverage.options');

        if (['foo' => 'test'] !== $options['output']) {
            throw new Exception('Default format is not singular output');
        }
    }
}


spec/Listener/CodeCoverageListenerSpec.php
<?php

declare(strict_types=1);

namespace spec\FriendsOfPhpSpec\PhpSpec\CodeCoverage\Listener;

use PhpSpec\Console\ConsoleIO;
use PhpSpec\Event\SuiteEvent;
use PhpSpec\ObjectBehavior;
use Prophecy\Argument;
use SebastianBergmann\CodeCoverage\CodeCoverage;
use SebastianBergmann\CodeCoverage\Filter;
use SebastianBergmann\CodeCoverage\Report;

/**
 * @author Henrik Bjornskov
 */
class CodeCoverageListenerSpec extends ObjectBehavior
{
    /**
     * Disabled due to tests breaking as php-code-coverage marked their classes
     * final and we cannot mock them. The tests should be converted into proper
     * functional (integration) tests instead. This file is left for reference.
     *
     * @see https://github.com/leanphp/phpspec-code-coverage/issues/19
     *
     * function let(ConsoleIO $io, CodeCoverage $coverage)
     * {
     * $this->beConstructedWith($io, $coverage, array());
     * }
     *
     * function it_is_initializable()
     * {
     * $this->shouldHaveType('LeanPHP\PhpSpec\CodeCoverage\Listener\CodeCoverageListener');
     * }
     *
     * function it_should_run_all_reports(
     * CodeCoverage $coverage,
     * Report\Clover $clover,
     * Report\PHP $php,
     * SuiteEvent $event,
     * ConsoleIO $io
     * ) {
     * $reports = array(
     * 'clover' => $clover,
     * 'php' =>  $php
     * );
     *
     * $io->isVerbose()->willReturn(false);
     *
     * $this->beConstructedWith($io, $coverage, $reports);
     * $this->setOptions(array(
     * 'format' => array('clover', 'php'),
     * 'output' => array(
     * 'clover' => 'coverage.xml',
     * 'php' => 'coverage.php'
     * )
     * ));
     *
     * $clover->process($coverage, 'coverage.xml')->shouldBeCalled();
     * $php->process($coverage, 'coverage.php')->shouldBeCalled();
     *
     * $this->afterSuite($event);
     * }
     *
     * function it_should_color_output_text_report_by_default(
     * CodeCoverage $coverage,
     * Report\Text $text,
     * SuiteEvent $event,
     * ConsoleIO $io
     * ) {
     * $reports = array(
     * 'text' => $text
     * );
     *
     * $this->beConstructedWith($io, $coverage, $reports);
     * $this->setOptions(array(
     * 'format' => 'text'
     * ));
     *
     * $io->isVerbose()->willReturn(false);
     * $io->isDecorated()->willReturn(true);
     *
     * $text->process($coverage, true)->willReturn('report');
     * $io->writeln('report')->shouldBeCalled();
     *
     * $this->afterSuite($event);
     * }
     *
     * function it_should_not_color_output_text_report_unless_specified(
     * CodeCoverage $coverage,
     * Report\Text $text,
     * SuiteEvent $event,
     * ConsoleIO $io
     * ) {
     * $reports = array(
     * 'text' => $text
     * );
     *
     * $this->beConstructedWith($io, $coverage, $reports);
     * $this->setOptions(array(
     * 'format' => 'text'
     * ));
     *
     * $io->isVerbose()->willReturn(false);
     * $io->isDecorated()->willReturn(false);
     *
     * $text->process($coverage, false)->willReturn('report');
     * $io->writeln('report')->shouldBeCalled();
     *
     * $this->afterSuite($event);
     * }
     *
     * function it_should_output_html_report(
     * CodeCoverage $coverage,
     * Report\Html\Facade $html,
     * SuiteEvent $event,
     * ConsoleIO $io
     * ) {
     * $reports = array(
     * 'html' => $html
     * );
     *
     * $this->beConstructedWith($io, $coverage, $reports);
     * $this->setOptions(array(
     * 'format' => 'html',
     * 'output' => array('html' => 'coverage'),
     * ));
     *
     * $io->isVerbose()->willReturn(false);
     * $io->writeln(Argument::any())->shouldNotBeCalled();
     *
     * $html->process($coverage, 'coverage')->willReturn('report');
     *
     * $this->afterSuite($event);
     * }
     *
     * function it_should_provide_extra_output_in_verbose_mode(
     * CodeCoverage $coverage,
     * Report\Html\Facade $html,
     * SuiteEvent $event,
     * ConsoleIO $io
     * ) {
     * $reports = array(
     * 'html' => $html,
     * );
     *
     * $this->beConstructedWith($io, $coverage, $reports);
     * $this->setOptions(array(
     * 'format' => 'html',
     * 'output' => array('html' => 'coverage'),
     * ));
     *
     * $io->isVerbose()->willReturn(true);
     * $io->writeln('')->shouldBeCalled();
     * $io->writeln('Generating code coverage report in html format ...')->shouldBeCalled();
     *
     * $this->afterSuite($event);
     * }
     *
     * function it_should_correctly_handle_black_listed_files_and_directories(
     * CodeCoverage $coverage,
     * SuiteEvent $event,
     * Filter $filter,
     * ConsoleIO $io
     * )
     * {
     * $this->beConstructedWith($io, $coverage, array());
     *
     * $coverage->filter()->willReturn($filter);
     *
     * $this->setOptions(array(
     * 'whitelist' => array('src'),
     * 'blacklist' => array('src/filter'),
     * 'whitelist_files' => array('src/filter/whilelisted_file'),
     * 'blacklist_files' => array('src/filtered_file')
     * ));
     *
     * $filter->addDirectoryToWhitelist('src')->shouldBeCalled();
     * $filter->removeDirectoryFromWhitelist('src/filter')->shouldBeCalled();
     * $filter->addFileToWhitelist('src/filter/whilelisted_file')->shouldBeCalled();
     * $filter->removeFileFromWhitelist('src/filtered_file')->shouldBeCalled();
     *
     * $this->beforeSuite($event);
     * }
     */
}


spec/Exception/NoCoverageDriverAvailableExceptionSpec.php
<?php

declare(strict_types=1);

namespace spec\FriendsOfPhpSpec\PhpSpec\CodeCoverage\Exception;

use FriendsOfPhpSpec\PhpSpec\CodeCoverage\Exception\NoCoverageDriverAvailableException;
use PhpSpec\ObjectBehavior;
use RuntimeException;

/**
 * @author Stéphane Hulard
 */
class NoCoverageDriverAvailableExceptionSpec extends ObjectBehavior
{
    public function it_is_initializable(): void
    {
        $this->shouldBeAnInstanceOf(RuntimeException::class);
        $this->shouldHaveType(NoCoverageDriverAvailableException::class);
    }
}


src/Listener/CodeCoverageListener.php
<?php

/**
 * This file is part of the friends-of-phpspec/phpspec-code-coverage package.
 *
 * @author  ek9 <dev@ek9.co>
 * @license MIT
 *
 * For the full copyright and license information, please see the LICENSE file
 * that was distributed with this source code.
 */

declare(strict_types=1);

namespace FriendsOfPhpSpec\PhpSpec\CodeCoverage\Listener;

use PhpSpec\Console\ConsoleIO;
use PhpSpec\Event\ExampleEvent;
use PhpSpec\Event\SuiteEvent;
use SebastianBergmann\CodeCoverage\CodeCoverage;
use SebastianBergmann\CodeCoverage\Report;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;

/**
 * @author Henrik Bjornskov
 */
class CodeCoverageListener implements EventSubscriberInterface
{
    private $coverage;

    private $io;

    private $options;

    private $reports;

    private $skipCoverage;

    public function __construct(ConsoleIO $io, CodeCoverage $coverage, array $reports, bool $skipCoverage = false)
    {
        $this->io = $io;
        $this->coverage = $coverage;
        $this->reports = $reports;
        $this->options = [
            'whitelist' => ['src', 'lib'],
            'blacklist' => ['test', 'vendor', 'spec'],
            'whitelist_files' => [],
            'blacklist_files' => [],
            'output' => ['html' => 'coverage'],
            'format' => ['html'],
        ];

        $this->skipCoverage = $skipCoverage;
    }

    public function afterExample(ExampleEvent $event): void
    {
        if ($this->skipCoverage) {
            return;
        }

        $this->coverage->stop();
    }

    public function afterSuite(SuiteEvent $event): void
    {
        if ($this->skipCoverage) {
            if ($this->io && $this->io->isVerbose()) {
                $this->io->writeln('Skipping code coverage generation');
            }

            return;
        }

        if ($this->io && $this->io->isVerbose()) {
            $this->io->writeln();
        }

        foreach ($this->reports as $format => $report) {
            if ($this->io && $this->io->isVerbose()) {
                $this->io->writeln(sprintf('Generating code coverage report in %s format ...', $format));
            }

            if ($report instanceof Report\Text) {
                $this->io->writeln(
                    $report->process($this->coverage, $this->io->isDecorated())
                );
            } else {
                $report->process($this->coverage, $this->options['output'][$format]);
            }
        }
    }

    public function beforeExample(ExampleEvent $event): void
    {
        if ($this->skipCoverage) {
            return;
        }

        $example = $event->getExample();

        $name = null;

        if (null !== $spec = $example->getSpecification()) {
            $name = $spec->getClassReflection()->getName();
        }

        $name = strtr('%spec%::%example%', [
            '%spec%' => $name,
            '%example%' => $example->getFunctionReflection()->getName(),
        ]);

        $this->coverage->start($name);
    }

    /**
     * Note: We use array_map() instead of array_walk() because the latter expects
     * the callback to take the value as the first and the index as the seconds parameter.
     *
     * @param SuiteEvent $event
     */
    public function beforeSuite(SuiteEvent $event): void
    {
        if ($this->skipCoverage) {
            return;
        }

        $filter = $this->coverage->filter();

        foreach ($this->options['whitelist'] as $option) {
            $filter->addDirectoryToWhitelist($option);
        }

        foreach ($this->options['blacklist'] as $option) {
            $filter->removeDirectoryFromWhitelist($option);
        }

        foreach ($this->options['whitelist_files'] as $option) {
            $filter->addFilesToWhitelist($option);
        }

        foreach ($this->options['blacklist_files'] as $option) {
            $filter->removeFileFromWhitelist($option);
        }
    }

    /**
     * {@inheritdoc}
     */
    public static function getSubscribedEvents(): array
    {
        return [
            'beforeExample' => ['beforeExample', -10],
            'afterExample' => ['afterExample', -10],
            'beforeSuite' => ['beforeSuite', -10],
            'afterSuite' => ['afterSuite', -10],
        ];
    }

    public function setOptions(array $options): void
    {
        $this->options = $options + $this->options;
    }
}


src/Listener/NoCodeCoverageListener.php
<?php

namespace LeanPHP\PhpSpec\CodeCoverage\Listener;

use PhpSpec\Console\ConsoleIO;
use PhpSpec\Event\SuiteEvent;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;

/**
 * @author Danny Lewis
 */
class NoCodeCoverageListener implements EventSubscriberInterface
{
    private $io;

    public function __construct(ConsoleIO $io)
    {
        $this->io = $io;
    }

    public function showDisabledMessage(SuiteEvent $event)
    {
        $this->io->writeln('Code coverage is disabled, enable with "coverage" flag. e.g. bin/phpspec run --coverage');
    }

    /**
     * {@inheritDoc}
     */
    public static function getSubscribedEvents()
    {
        return array(
            'beforeSuite'   => array('showDisabledMessage', -10),
            'afterSuite'    => array('showDisabledMessage', -10),
        );
    }
}


src/CodeCoverageExtension.php
<?php

/**
 * This file is part of the friends-of-phpspec/phpspec-code-coverage package.
 *
 * @author ek9 <dev@ek9.co>
 * @license MIT
 *
 * For the full copyright and license information, please see the LICENSE file
 * that was distributed with this source code.
 */

declare(strict_types=1);

namespace FriendsOfPhpSpec\PhpSpec\CodeCoverage;

use FriendsOfPhpSpec\PhpSpec\CodeCoverage\Exception\NoCoverageDriverAvailableException;
use FriendsOfPhpSpec\PhpSpec\CodeCoverage\Listener\CodeCoverageListener;
use PhpSpec\Extension;
use PhpSpec\ServiceContainer;
use RuntimeException;
use SebastianBergmann\CodeCoverage\CodeCoverage;
use SebastianBergmann\CodeCoverage\Filter;
use SebastianBergmann\CodeCoverage\Report;
use SebastianBergmann\CodeCoverage\Version;
use Symfony\Component\Console\Input\InputOption;

use function count;
use function is_array;

/**
 * Injects Code Coverage Event Subscriber into the EventDispatcher.
 * The Subscriber will add Code Coverage information before each example.
 *
 * @author Henrik Bjornskov
 */
class CodeCoverageExtension implements Extension
{
    /**
     * {@inheritdoc}
     */
    public function load(ServiceContainer $container, array $params = []): void
    {
        foreach ($container->getByTag('console.commands') as $command) {
            $command->addOption('no-coverage', null, InputOption::VALUE_NONE, 'Skip code coverage generation');
        }

        $container->define('code_coverage.filter', static function () {
            return new Filter();
        });

        $container->define('code_coverage', static function ($container) {
            try {
                $coverage = new CodeCoverage(null, $container->get('code_coverage.filter'));
            } catch (RuntimeException $error) {
                throw new NoCoverageDriverAvailableException(
                    'There is no available coverage driver to be used.',
                    0,
                    $error
                );
            }

            return $coverage;
        });

        $container->define('code_coverage.options', static function ($container) use ($params) {
            $options = !empty($params) ? $params : $container->getParam('code_coverage');

            if (!isset($options['format'])) {
                $options['format'] = ['html'];
            } elseif (!is_array($options['format'])) {
                $options['format'] = (array) $options['format'];
            }

            if (isset($options['output'])) {
                if (!is_array($options['output']) && 1 === count($options['format'])) {
                    $format = $options['format'][0];
                    $options['output'] = [$format => $options['output']];
                }
            }

            if (!isset($options['show_uncovered_files'])) {
                $options['show_uncovered_files'] = true;
            }

            if (!isset($options['lower_upper_bound'])) {
                $options['lower_upper_bound'] = 35;
            }

            if (!isset($options['high_lower_bound'])) {
                $options['high_lower_bound'] = 70;
            }

            return $options;
        });

        $container->define('code_coverage.reports', static function ($container) {
            $options = $container->get('code_coverage.options');

            $reports = [];

            foreach ($options['format'] as $format) {
                switch ($format) {
                    case 'clover':
                        $reports['clover'] = new Report\Clover();

                        break;
                    case 'php':
                        $reports['php'] = new Report\PHP();

                        break;
                    case 'text':
                        $reports['text'] = new Report\Text(
                            $options['lower_upper_bound'],
                            $options['high_lower_bound'],
                            $options['show_uncovered_files'],
                            /* $showOnlySummary */
                            false
                        );

                        break;
                    case 'xml':
                        $reports['xml'] = new Report\Xml\Facade(Version::id());

                        break;
                    case 'crap4j':
                        $reports['crap4j'] = new Report\Crap4j();

                        break;
                    case 'html':
                        $reports['html'] = new Report\Html\Facade();

                        break;
                }
            }

            $container->setParam('code_coverage', $options);

            return $reports;
        });

        $container->define('event_dispatcher.listeners.code_coverage', static function ($container) {
            $skipCoverage = false;
            $input = $container->get('console.input');

            if ($input->hasOption('no-coverage') && $input->getOption('no-coverage')) {
                $skipCoverage = true;
            }

            $listener = new CodeCoverageListener(
                $container->get('console.io'),
                $container->get('code_coverage'),
                $container->get('code_coverage.reports'),
                $skipCoverage
            );
            $listener->setOptions($container->getParam('code_coverage', []));

            return $listener;
        }, ['event_dispatcher.listeners']);
    }
}
